src/test/datascience/data-viewing/dataViewerDependencyService.unit.test.ts

```diff
@@ -35,87 +59,347 @@ suite('DataScience - DataViewerDependencyService', () => {
             sysVersion: '',
             version: new SemVer('3.3.3')
         };
-        pythonExecService = mock<IPythonExecutionService>();
-        installer = mock(ProductInstaller);
         appShell = mock(ApplicationShell);
-        pythonExecFactory = mock(PythonExecutionFactory);
-        interpreterService = mock<IInterpreterService>();
+        kernelProvider = instance(mock<IKernelProvider>());
+        kernel = instance(mock<IKernel>());
+        sessionExecutionContext = [];
+
+        dependencyService = new DataViewerDependencyService(instance(appShell), kernelProvider, false);
+    });
+
+    test('What happens if there are no idle kernels?', async () => {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (kernel.session as any) = jupyterSession;
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (kernel.status as any) = 'busy';
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (kernelProvider.kernels as any) = [kernel];
+
+        shellFuture = {
```

Comment:
I might be wrong but I assume that plain functions are not very easy to mock? One would need to change when they are being loaded.

Reply:
Do a search for sinon.stub. It's actually very easy to mock them.

Reply:
I don't think we can stub them, it works only if the stubbing is applied before the module is imported.
However since the imports are at the top level it might not work (as all modules end up getting loaded & imported before the tests run)

Reply:
Yes, but that really depends on how the module is built. Plain functions can’t be stubbed. That doesn’t mean it can’t work in the case of executeSilently, I haven’t taken the time yet.

Reply:
Ok it worked! yay 🙂 had to change imports, but not in a bad way. Here's the commit: https://github.com/microsoft/vscode-jupyter/pull/10545/commits/ae2d66776f9e3984c55e79cd5d032061af7a1c19